sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -141,19 +141,25 @@ public static PipelineOptions createPipelineOptions(Map<String, String> map) {
   }
 
   static List<Row> toRowList(PipelineOptions options, BeamRelNode node) {
+    checkAndFailOnNonDirectRunner(options, "toRowList is only available in DirectRunner.");
+
     if (node instanceof BeamIOSinkRel) {
       throw new UnsupportedOperationException("Does not support BeamIOSinkRel in toRowList.");
     } else if (isLimitQuery(node)) {
-      throw new UnsupportedOperationException("Does not support queries with LIMIT in toRowList.");
+      return limitCollectRows(options, node).stream().collect(Collectors.toList());
     }
+
     return collectRows(options, node).stream().collect(Collectors.toList());
   }
 
   static Enumerable<Object> toEnumerable(PipelineOptions options, BeamRelNode node) {
+    checkAndFailOnNonDirectRunner(
+        options, "SELECT without INSERT is only supported in DirectRunner in SQL Shell.");
+
     if (node instanceof BeamIOSinkRel) {
       return count(options, node);
     } else if (isLimitQuery(node)) {
-      return limitCollect(options, node);
+      return Linq4j.asEnumerable(rowToAvaticaAndUnboxValues(limitCollectRows(options, node)));
```

Comment:
Looking at this I think it is possible to refactor this even further. `toRowList()` and `toEnumerable()` are essentially the same thing now with minor differences. Would it be possible to do the following?

* add support for count to `toRowList()` as mentioned in another comment;
* don't call `limitCollectRows()` here, but call `toRowList()` instead;
* this would probably make this method a one-line call to `Linq4j.asEnumerable(rowToAvaticaAndUnboxValues..`

Reply:
limit collect relies on `limitRun`, which requires cancel pipeline based on collected value size. That's the reason limit collect and collect are in different code path.

Reply:
I am not sure I understand. I am not proposing to re-use the code between `limitCollect` and `collect` parts. What I mean is:

Change `toRowList()` to something like:

```
static List<Row> toRowList(PipelineOptions options, BeamRelNode node) {
    Collection<Row> rows;
    if (node instanceof BeamIOSinkRel) {
     rows = count();
    } else if (isLimitQuery(node)) {
      rows = limitCollectRows(options, node);
    } else {
      rows = collectRows(options, node);
    }
    return new ArrayList<>(rows);
  }
```

Then  `toEnumerable` may become something like this:

```
static Enumerable<Object> toEnumerable(PipelineOptions options, BeamRelNode node) {
    Collection<Row> rows = toRowList(options, node);
    return Linq4j.asEnumerable(rowToAvaticaAndUnboxValues(rows));
}
```

I may be missing something though.

Reply:
I see.

What you are missing is `count` does not collect rows: it uses metrics API to count (https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java#L370).

Reply:
It might be feasible to re-implement count to make it return a `List<Row>{ROW<int>(count)}`, but it seems over-kill for achieving the goal of merging toRowList and toEnumerable.

Reply:
It does not matter how `count()` calculates the results, it can return a constant or a random result or call some service in the cloud. It's a function in a relational programming model, so the result is expected to be a relation. Sure, it feels not necessary to wrap a single value into a row so that it can be immediately unwrapped (in one case). But my argument is:
 * it is natural to define a function in such environment as returning Rows, so it will not be unexpected to the readers;
 * it is harder to read two duplicate versions of semantically the same code and figure out the differences, then to understand why we have to wrap a single value in a row;
 * no special casing;
 * it is less code, less things to read, break and test;